revocation/internal/crl/crl.go

```diff
@@ -0,0 +1,323 @@
+				Error: fmt.Errorf("CRL is not supported in the certificate %s", cert.Subject.CommonName),
+			}},
+		}
+	}
+
+	if opts.HTTPClient == nil {
+		return &result.CertRevocationResult{
+			Result: result.ResultUnknown,
+			CRLResults: []*result.CRLResult{{
+				Error: errors.New("HTTP client is nil"),
+			}},
+		}
+	}
+
+	// The CRLDistributionPoints contains the URIs of all the CRL distribution
+	// points. Since it does not distinguish the reason field, it needs to check
+	// all the URIs to avoid missing any partial CRLs.
+	//
+	// For the majority of the certificates, there is only one CRL distribution
+	// point with one CRL URI, which will be cached, so checking all the URIs is
+	// not a performance issue.
+	var (
+		crlResults []*result.CRLResult
+		lastErr    error
+		crlURL     string
+	)
+	for _, crlURL = range cert.CRLDistributionPoints {
+		baseCRL, err := download(ctx, crlURL, opts.HTTPClient)
+		if err != nil {
+			lastErr = fmt.Errorf("failed to download CRL from %s: %w", crlURL, err)
+			break
+		}
+
+		if err = validate(baseCRL, issuer); err != nil {
+			lastErr = fmt.Errorf("failed to validate CRL from %s: %w", crlURL, err)
+			break
+		}
+
+		crlResult, err := checkRevocation(cert, baseCRL, opts.SigningTime, crlURL)
+		if err != nil {
+			lastErr = fmt.Errorf("failed to check revocation status from %s: %w", crlURL, err)
+			break
+		}
+		if hasDeltaCRL(baseCRL) {
+			// added an error to indicate that delta CRLs are not checked
+			crlResult.Error = ErrDeltaCRLNotChecked
+		}
+		crlResults = append(crlResults, crlResult)
+
+		if crlResult.Result == result.ResultRevoked {
+			return &result.CertRevocationResult{
+				Result:     result.ResultRevoked,
+				CRLResults: crlResults,
+			}
+		}
+	}
+
+	if lastErr != nil {
+		crlResults = append(crlResults, &result.CRLResult{
+			Result: result.ResultUnknown,
+			URI:    crlURL,
+			Error:  lastErr,
+		})
+		return &result.CertRevocationResult{
+			Result:     result.ResultUnknown,
+			CRLResults: crlResults,
+		}
+	}
+
+	return &result.CertRevocationResult{
+		Result:     result.ResultOK,
+		CRLResults: crlResults,
+	}
+}
+
+// Supported checks if the certificate supports CRL.
+func Supported(cert *x509.Certificate) bool {
+	return cert != nil && len(cert.CRLDistributionPoints) > 0
+}
+
+func validate(crl *x509.RevocationList, issuer *x509.Certificate) error {
+	// check signature
+	if err := crl.CheckSignatureFrom(issuer); err != nil {
+		return fmt.Errorf("CRL signature verification failed: %w", err)
+	}
+
+	// check validity
+	if !crl.NextUpdate.IsZero() && time.Now().After(crl.NextUpdate) {
+		return fmt.Errorf("CRL is expired: %v", crl.NextUpdate)
+	}
+
+	for _, ext := range crl.Extensions {
+		if ext.Id.Equal(oidIssuingDistributionPoint) {
+			// IssuingDistributionPoint is a critical extension that identifies
+			// the scope of the CRL. Since we will check all the CRL
+			// distribution points, it is not necessary to check this extension.
+			continue
+		}
```

Comment:
According to [RFC 5280 5.2](https://www.rfc-editor.org/rfc/rfc5280#section-5.2):
Conforming CRL issuers are REQUIRED to include the authority key identifier ([Section 5.2.1](https://www.rfc-editor.org/rfc/rfc5280#section-5.2.1)) and the CRL number ([Section 5.2.3](https://www.rfc-editor.org/rfc/rfc5280#section-5.2.3)) extensions in all CRLs issued.

Should we add validations of these two fields here?

Reply:
Yes, the validation is mainly about checking the existence. CRL that missing these fields should be considered invalid and rejected by Notation.

Reply:
https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/crypto/x509/parser.go;l=1214-1225

The Go built-in `ParseRevocationList` function handles those two extensions, but the library doesn't enforce their existence. I think we should keep it as is without further restrictions, which is the same approach that the Go library takes.